src/Servers/Kestrel/Transport.Quic/src/Internal/QuicConnectionContext.cs

```diff
@@ -111,22 +127,46 @@ public override void Abort(ConnectionAbortedException abortReason)
             catch (QuicConnectionAbortedException ex)
             {
                 // Shutdown initiated by peer, abortive.
-                _log.ConnectionAborted(this, ex);
+                Error = ex.ErrorCode;
+                _log.ConnectionAborted(this, ex.ErrorCode, ex);
 
                 ThreadPool.UnsafeQueueUserWorkItem(state =>
                 {
                     state.CancelConnectionClosedToken();
                 },
                 this,
                 preferLocal: false);
+
+                // Throw error so consumer sees the connection is aborted by peer.
+                throw new ConnectionResetException(ex.Message, ex);
             }
             catch (QuicOperationAbortedException)
             {
-                // Shutdown initiated by us
+                // Shutdown initiated by us.
 
-                // Allow for graceful closure.
+                lock (_shutdownLock)
+                {
+                    // Connection has been aborted. Throw reason exception.
+                    _abortReason?.Throw();
+                }
+            }
+            catch (OperationCanceledException)
+            {
+                // Shutdown initiated by us.
+
+                lock (_shutdownLock)
+                {
+                    // Connection has been aborted. Throw reason exception.
+                    _abortReason?.Throw();
+                }
+            }
+            catch (Exception ex)
+            {
+                _log.LogWarning(0, ex, $"Unexpected exception in {nameof(QuicConnectionContext)}.{nameof(AcceptAsync)}.");
```

Comment:
Please don't log and re-throw, especially at such a high level, it cause duplicate logs. It looks like you intend this more as an Assert anyways.